scripts/skills/perks/perk_rf_professional.nut

```diff
@@ -15,45 +15,85 @@ this.perk_rf_professional <- ::inherit("scripts/skills/skill", {
-			o.m.IsSerialized = false;
-			o.m.IsRefundable = false;
-		}));
-		this.getContainer().add(::MSU.new("scripts/skills/perks/perk_rf_formidable_approach", function(o) {
-			o.m.IsSerialized = false;
-			o.m.IsRefundable = false;
-		}));
-		this.getContainer().add(::MSU.new("scripts/skills/perks/perk_duelist", function(o) {
-			o.m.IsSerialized = false;
-			o.m.IsRefundable = false;
-		}));
-		this.getContainer().add(::MSU.new("scripts/skills/perks/perk_rf_weapon_master", function(o) {
-			o.m.IsSerialized = false;
-			o.m.IsRefundable = false;
-		}));
-
-		if (this.m.IsNew)
+		local chosenPerkGroupID;
+
+		local actor = this.getContainer().getActor();
+		if (actor.getFlags().has("RF_ProfessionalPerkGroupID"))
+		{
+			chosenPerkGroupID = actor.getFlags().get("RF_ProfessionalPerkGroupID");
+		}
+		else
+		{
+			local perkGroups = [
+				"pg.rf_axe",
+				"pg.rf_cleaver",
+				"pg.rf_dagger",
+				"pg.rf_hammer",
+				"pg.rf_mace",
+				"pg.rf_polearm",
+				"pg.rf_spear",
+				"pg.rf_sword"
+			];
+
+			local options = [];
+			foreach (pg in perkGroups)
+			{
+				if (actor.getPerkTree().hasPerkGroup(pg))
+				{
+					options.push(pg);
+				}
+			}
+
+			if (options.len() == 0)
+				return;
+
+			chosenPerkGroupID = ::MSU.Array.rand(options);
+
+			actor.getFlags().set("RF_ProfessionalPerkGroupID", chosenPerkGroupID);
+		}
+
+		local perkIDs = [];
+		foreach (row in ::DynamicPerks.PerkGroups.findById(chosenPerkGroupID).getTree())
 		{
-			local perkTree = this.getContainer().getActor().getPerkTree();
+			foreach (perkID in row)
+			{
+				perkIDs.push(perkID);
+			}
+		}
+
+		if (perkIDs.len() > 2)
+			perkIDs = perkIDs.slice(0, 2);
 
-			perkTree.addPerk("perk.rf_death_dealer", 2);
-			perkTree.addPerk("perk.shield_expert", 3);
-			perkTree.addPerk("perk.rf_formidable_approach", 3);
-			perkTree.addPerk("perk.duelist", 6);
-			perkTree.addPerk("perk.rf_weapon_master", 7);
+		foreach (id in perkIDs)
+		{
+			this.getContainer().add(::MSU.new(actor.getPerkTree().getPerk(id).Script, function(o) {
+				o.m.IsSerialized = false;
+				o.m.IsRefundable = false;
+			}));
 		}
 	}
 
 	function onRemoved()
 	{
-		this.getContainer().removeByStackByID("perk.rf_death_dealer", false);
-		this.getContainer().removeByStackByID("perk.shield_expert", false);
-		this.getContainer().removeByStackByID("perk.rf_formidable_approach", false);
-		this.getContainer().removeByStackByID("perk.duelist", false);
-		this.getContainer().removeByStackByID("perk.rf_weapon_master", false);
+		local actor = this.getContainer().getActor();
+		local chosenPerkGroupID = actor.getFlags().get("RF_ProfessionalPerkGroupID");
+		actor.getFlags().remove("RF_ProfessionalPerkGroupID");
+
+		local perkIDs = [];
+		foreach (row in ::DynamicPerks.PerkGroups.findById(chosenPerkGroupID).getTree())
+		{
+			foreach (perkID in row)
+			{
+				perkIDs.push(perkID);
+			}
+		}
+
+		if (perkIDs.len() > 2)
+			perkIDs = perkIDs.slice(0, 2);
```

Comment:
You always remove the first two perks in that perk group. But those were not necessarily all added by `Professional`, if a player learned those perks before they learned `Professional`.

This is a minor edge case.

Maybe we prevent `Professional` from being picked if the player has learned any perk from any weapon tree.

Reply:
The perks are added and removed via stack (`mod_stack_based_skills`) so it should be fine. Even if the bro already had the perk before picking Professional, Professional will add a "stack" of that perk, and when removed will only remove that stack. Meanwhile, if the bro "unlearns" that perk, he will unlearn the original perk, but the stack based addition from Professional will remain.

An issue here can happen with regards to refundability of the perks. So the original perk was refundable, but the one added by Professional isn't - however, the way stack based skills mod works is that the data is kept from the serialized version of the skill. So after the original perk is unlearned, the one added by Professional will also be refundable. I think a meaningful solution to this should be that all non-serialized skills should be considered not-refundable by default. I can update Dynamic Perks Framework to account for this.

Reply:
Ok, I made simple issue over there for a later time.
No further adjustments are needed regarding potential edge cases.